skglm/estimators.py

```diff
@@ -1,31 +1,164 @@
+from skglm.solvers.cd_solver import cd_solver
+from skglm.solvers.multitask_bcd_solver import multitask_bcd_solver
+from skglm.datafits import Quadratic, Logistic, QuadraticSVC, QuadraticMultiTask
+from skglm.penalties import L1, WeightedL1, L1_plus_L2, MCPenalty, IndicatorBox, L2_1
+
+
+def _glm_fit(X, y, model, datafit, penalty):
+    is_classif = False
+    if isinstance(datafit, Logistic) or isinstance(datafit, QuadraticSVC):
+        is_classif = True
+
+    if is_classif:
+        check_classification_targets(y)
+        enc = LabelEncoder()
+        y = enc.fit_transform(y)
+        model.classes_ = enc.classes_
+        n_classes_ = len(model.classes_)
+        is_sparse = issparse(X)
+        if n_classes_ <= 2:
+            y = 2 * y - 1
+        X = check_array(
+            X, accept_sparse="csc", dtype=np.float64, accept_large_sparse=False)
+        y = check_array(
+            y, ensure_2d=False, dtype=X.dtype.type, accept_large_sparse=False)
+        check_consistent_length(X, y)
+    else:
+        check_X_params = dict(
+            dtype=[np.float64, np.float32], order='F',
+            accept_sparse='csc', copy=model.fit_intercept)
+        check_y_params = dict(ensure_2d=False, order='F')
+
+        X, y = model._validate_data(
+            X, y, validate_separately=(check_X_params, check_y_params))
+        X = check_array(X, 'csc', dtype=[np.float64, np.float32],
+                        order='F', copy=False, accept_large_sparse=False)
+        y = check_array(y, 'csc', dtype=X.dtype.type, order='F', copy=False,
+                        ensure_2d=False)
+
+    if y.ndim == 2 and y.shape[1] == 1:
+        warnings.warn("DataConversionWarning('A column-vector y"
+                      " was passed when a 1d array was expected")
+        y = y[:, 0]
+
+    if not hasattr(model, "n_features_in_"):
+        model.n_features_in_ = X.shape[1]
+
+    n_samples = X.shape[0]
+    if n_samples != y.shape[0]:
+        raise ValueError("X and y have inconsistent dimensions (%d != %d)"
+                         % (n_samples, y.shape[0]))
+
+    if not model.warm_start or not hasattr(model, "coef_"):
+        model.coef_ = None
+
+    if is_classif and n_classes_ > 2:
+        model.coef_ = np.empty([len(model.classes_), X.shape[1]])
+        if isinstance(datafit, QuadraticSVC):
+            model.dual_coef_ = np.empty([len(model.classes_), X.shape[0]])
+        model.intercept_ = 0
+        multiclass = OneVsRestClassifier(model).fit(X, y)
+        model.coef_ = np.array(
+            [clf.coef_[0] for clf in multiclass.estimators_])
+        if isinstance(datafit, QuadraticSVC):
+            model.dual_coef_ = np.array(
+                [clf.dual_coef_[0] for clf in multiclass.estimators_])
+        model.n_iter_ = max(
+            clf.n_iter_ for clf in multiclass.estimators_)
+        return model
+    if is_classif and n_classes_ <= 2 and isinstance(datafit, QuadraticSVC):
```

Comment:
cosmetic: when two if blocks are unrelated (not like an if elif), it often helps to put a blank line between them